lib/Sema/ConstraintSystem.cpp

```diff
@@ -1506,9 +1507,42 @@ static Type getEffectiveOverloadType(const OverloadChoice &overload) {
                              genericFn->getExtInfo());
   }
 
-  // If this declaration is within a type context, bail out.
+  // If this declaration is within a type context, we might not be able
+  // to handle it.
   if (decl->getDeclContext()->isTypeContext()) {
-    return Type();
+    if (!allowMembers)
+      return Type();
+
+    if (auto subscript = dyn_cast<SubscriptDecl>(decl)) {
+      auto elementTy = subscript->getElementInterfaceType();
+
+      if (doesStorageProduceLValue(subscript, overload.getBaseType(), useDC))
+        elementTy = LValueType::get(elementTy);
+
+      // See ConstraintSystem::resolveOverload() -- optional and dynamic
+      // subscripts are a special case, because the optionality is
+      // applied to the result type and not the type of the reference.
+      if (subscript->getAttrs().hasAttribute<OptionalAttr>() ||
+          overload.getKind() == OverloadChoiceKind::DeclViaDynamic)
+        elementTy = OptionalType::get(elementTy->getRValueType());
+
+      auto indices = subscript->getInterfaceType()
+                       ->castTo<AnyFunctionType>()->getParams();
+      type = FunctionType::get(indices, elementTy);
+    } else if (auto var = dyn_cast<VarDecl>(decl)) {
+      type = var->getInterfaceType();
+      if (doesStorageProduceLValue(var, overload.getBaseType(), useDC))
+        type = LValueType::get(type);
+    } else if (isa<FuncDecl>(decl) || isa<EnumElementDecl>(decl)) {
```

Comment:
What about constructors?

Reply:
I've pulled the fix for constructors into https://github.com/apple/swift/pull/23012, thanks!